keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,388 @@
+
+Operator developer experience, then, focuses on customizing diagnostics
+information to gather information about the running application. The following
+forms are available, subject to change over time:
+
+- **Copy**: Copy a file out of a running pod. This is useful for non-stdout/err
+  logs, configuration files, and other artifacts generated by an application.
+  Higher level resources can also be used, which will copy the file on all pods
+  selected by that resource.
+- **Command**: Run a command on a running pod and copy the stdout/err. Higher level
+  resources can also be used, which will run the command on all pods selected by
+  that resource.
+- **Task**: Run a KUDO task and copy the stdout and other arbitrary files.
+- **HTTP**: Make an HTTP request from the KUDO controller manager to a named
+  service and port and copy the result of the request.
+
+While some of these are redundant (HTTP can be a command or job), the intent
+is to provide a high level experience where possible so that operator developers
+don't necessarily need to maintain a `curl` container as part of their
+application stack.
+
+Operator-defined diagnostics collection is defined in a new `diagnostics.bundle.resources`
+key in `operator.yaml`:
+
+```yaml
+diagnostics:
+  bundle:
+    resources:
+      - description: Zookeeper Configuration File
+        name: "zookeeper-configuration"
+        kind: Copy
+        spec:
+          path: 
+            - /opt/zookeeper/server.properties
+            - /opt/zookeeper/zoo.cfg
+          selector:
+            matchLabels:
+              app: zookeeper
+              heritage: kudo
+          objectRef:
+            kind: StatefulSet # Runs on ALL pods in the statefulset
+      - description: DNS information for running pod
+        name: "dns-information"
+        kind: Command
+        spec:
+          command: # Can be string or array
+            - nslookup
+            - google.com
+          selector:
+            matchLabels:
+              app: zookeeper
+              heritage: kudo
+          objectRef:
+            kind: Pod
+    filters:
+      - name: Authentication information
+        spec:
+          regex: "^host: %w+$"
+```
+
+This key is **OPTIONAL**. Default diagnostics collection will happen regardless
+of the `diagnostics.bundle` key's presence. Note, moving to a graph-based engine
+for KUDO will make selecting of resources much easier, rather than having to
+use magical strings with templates. Future iterations of this will reduce the
+complexity of selecting resources to run commands and files on.
+
+Filtering is an important part of diagnostics collection. It enables diagnostics
+to be portably sent to third parties that should not have sensitive information
+that logs and files can contain.
+
+By default, KUDO filters all resources (and custom resources) of values
+contained within the KUDO Instance's secrets. This is configurable with the
```

Comment:
Hm, this seems like a promise that is hard to deliver on... for example the secret may be a JSON file which contains a password (`{"user":"admin", "password": "foo"}`). How do we make KUDO recognize that it should elide that password when it encounters it in a log file in a different context (`... logging into 1.2.3.4 with password foo.`)?

Reply:
True. But we probably cannot avoid doing this anyway, otherwise the bundle cannot be safely shared. For files with known structure (e.g. yaml) it will be pretty doable. Logs will present a problem indeed, however we can just blindly black out whatever matches any secret as the initial approach, or there may be configurable phrase-based replacement. I may be wrong here, but I believe that placing sensitive information in the logs is a dubious practice after all